cognite/client/data_classes/datapoints.py

```diff
@@ -271,19 +333,69 @@ def __str__(self):
+            complete_granularity (str): When complete=True, what granularity should the index be given. Defaults to the minimum difference between timestamps.
+
         Returns:
             pandas.DataFrame: The datapoints list as a pandas DataFrame.
         """
         pd = utils._auxiliary.local_import("pandas")
-        dfs = [df.to_pandas(column_names=column_names) for df in self.data]
-        if dfs:
-            return pd.concat(dfs, axis="columns")
-        return pd.DataFrame()
+
+        complete = [s.strip() for s in (complete or "").split(",")]
+        if set(complete) - set(["fill", "dropna", ""]):
+            raise ValueError("complete should be 'fill', 'fill,dropna' or Falsy")
+        fill_flag = "fill" in complete
+        is_step_dict = {dp.id: False for dp in self.data}
+        complete_start = None
+        complete_end = None
+        if "fill" in complete:
+            interp_ids = [dp.id for dp in self.data if dp.interpolation is not None]
+            starts = [dp.timestamp[0] for dp in self.data if dp.timestamp]
+            ends = [dp.timestamp[-1] for dp in self.data if dp.timestamp]
+            complete_start = min(starts) if starts else None
+            complete_end = max(ends) if starts else None
+            if interp_ids:  # get isStep flag and pass it to completion to handle interpolation correctly
+                ts_meta = self._cognite_client.time_series.retrieve_multiple(ids=interp_ids)
+                is_step_dict.update({ts.id: bool(ts.is_step) for ts in ts_meta})
+        dfs = [
+            df.to_pandas(
+                column_names=column_names,
+                include_aggregate_name=include_aggregate_name,
+                complete=fill_flag,
+                complete_is_step=is_step_dict[df.id],
+                complete_granularity=complete_granularity,
+                complete_start=complete_start,
+                complete_end=complete_end,
+            )
+            for df in self.data
+        ]
+        if not dfs:
+            return pd.DataFrame()
+
+        df = pd.concat(dfs, axis="columns")
+        if "dropna" in complete:
+            supported_aggregates = ["sum", "count", "total_variance", "interpolation", "step_interpolation"]
+            aggregates_used = [ag for df in self.data for ag in df._get_non_empty_data_keys()]
+            not_supported = set(aggregates_used) - set(supported_aggregates + ["timestamp"])
+            if not_supported:
+                raise ValueError(
+                    "The aggregate(s) {} is not supported for dataframe completion with dropna, only {} are".format(
+                        [utils._auxiliary.to_camel_case(a) for a in not_supported],
+                        [utils._auxiliary.to_camel_case(a) for a in supported_aggregates],
+                    )
+                )
+            df.dropna(inplace=True)
+        return df
```

Comment:
Maybe try splitting this into a few methods for readability aswell?

Reply:
ok